src/Access/MemoryAccessStorage.cpp

```diff
@@ -201,6 +231,28 @@ bool MemoryAccessStorage::updateNoLock(const UUID & id, const UpdateFunc & updat
 }
 
 
+void MemoryAccessStorage::removeAllExcept(const std::vector<UUID> & ids_to_keep)
+{
+    std::lock_guard lock{mutex};
+    removeAllExceptNoLock(ids_to_keep);
+}
+
+void MemoryAccessStorage::removeAllExceptNoLock(const std::vector<UUID> & ids_to_keep)
+{
+    boost::container::flat_set<UUID> ids_to_keep_set{ids_to_keep.begin(), ids_to_keep.end()};
+    for (auto it = entries_by_id.begin(); it != entries_by_id.end();)
+    {
+        const auto & id = it->first;
+        ++it;
```

Comment:
```suggestion
    for (auto it = entries_by_id.begin(); it != entries_by_id.end(); ++it)
    {
        const auto & id = it->first;
```
why not like this?

Reply:
Because removeNoLock() invalidates any iterator pointing to a removed element. So we need to call `++it` before `removeNoLock()`.